relay-server/src/actors/store.rs

```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+    let mut client_config = ClientConfig::new();
+    match config
         .kafka_config(kafka_topic)
-        .context(ServerErrorKind::KafkaError)?;
+        .context(ServerErrorKind::KafkaError)?
+    {
+        KafkaConfig::Single(KafkaParams {
+            topic_name,
+            config_name,
+            params,
+        }) => {
+            if let Some(producer) = reused_producers.get(&config_name) {
+                return Ok(Producer::Single {
+                    topic_name: topic_name.to_string(),
+                    producer: Arc::clone(producer),
+                });
+            }
+
+            for config_p in params {
+                client_config.set(config_p.name.as_str(), config_p.value.as_str());
+            }
 
-    if let Some(producer) = reused_producers.get(&config_name) {
-        return Ok(Arc::clone(producer));
+            let producer = Arc::new(
+                client_config
+                    .create_with_context(CaptureErrorContext)
+                    .context(ServerErrorKind::KafkaError)?,
+            );
+
+            reused_producers.insert(config_name, Arc::clone(&producer));
+            Ok(Producer::Single {
+                topic_name: topic_name.to_string(),
+                producer,
+            })
+        }
+        KafkaConfig::Sharded { shards, configs } => {
+            let mut producers = BTreeMap::new();
+            for (shard, kafka_params) in configs {
+                if let Some(producer) = reused_producers.get(&kafka_params.config_name) {
+                    let cached_producer = Arc::clone(producer);
+                    producers.insert(
+                        shard,
+                        (kafka_params.topic_name.to_string(), cached_producer),
+                    );
+                    continue;
+                }
+                for config_p in kafka_params.params {
+                    client_config.set(config_p.name.as_str(), config_p.value.as_str());
+                }
+                let producer = Arc::new(
+                    client_config
+                        .create_with_context(CaptureErrorContext)
+                        .context(ServerErrorKind::KafkaError)?,
+                );
+                reused_producers.insert(kafka_params.config_name, Arc::clone(&producer));
+                producers.insert(shard, (kafka_params.topic_name.to_string(), producer));
+            }
+            Ok(Producer::Sharded(Sharded { shards, producers }))
+        }
     }
+}
 
-    let mut client_config = ClientConfig::new();
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
+    Single {
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded(Sharded),
+}
 
-    for config_p in kafka_config {
-        client_config.set(config_p.name.as_str(), config_p.value.as_str());
+/// Sharded producer configuration.
+pub struct Sharded {
```

Comment:
nit: Should we call this `ShardedProducer`?

Reply:
Yeah, this is must better name!  I will make a change.